File: generated_code/cpp/solution_25_2.cpp
#include <vector>
#include <algorithm>

// Definition for singly-linked list.
// Included for compilability as mandated by the instructions.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        // Base case: if list is empty or k is 1, no structural change is needed.
        if (!head || k == 1) return head;

        // Use a dummy node to simplify handling the new head of the list.
        ListNode dummy(0);
        dummy.next = head;

        ListNode* groupPrev = &dummy;
        ListNode* curr = head;

        // Count total number of nodes to handle groups efficiently.
        int count = 0;
        while (curr) {
            count++;
            curr = curr->next;
        }

        // Loop runs for every complete group of size k.
        // Time Complexity: O(N) since each node is visited at most twice.
        // Space Complexity: O(1) auxiliary space as we reverse in-place.
        while (count >= k) {
            curr = groupPrev->next;
            ListNode* nextNode = curr->next;

            // Standard in-place reversal for k-1 links within the current group.
            for (int i = 1; i < k; ++i) {
                curr->next = nextNode->next;
                nextNode->next = groupPrev->next;
                groupPrev->next = nextNode;
                nextNode = curr->next;
            }

            // Move groupPrev to the end of the reversed group (which is 'curr').
            groupPrev = curr;
            count -= k;
        }

        return dummy.next;
    }
};

File: generated_code/python/solution_25_2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Edge case: if k is 1 or list is empty, no reversal needed
        if not head or k == 1:
            return head
        
        # Dummy node to simplify handling the head of the modified list
        dummy = ListNode(0)
        dummy.next = head
        
        # Pointer to the node immediately before the group to be reversed
        group_prev = dummy
        
        while True:
            # Find the k-th node from group_prev
            kth = group_prev
            for _ in range(k):
                kth = kth.next
                if not kth:
                    break
            
            # If there are fewer than k nodes left, we are done
            if not kth:
                break
                
            # Track the node right after the current k-group
            group_next = kth.next
            
            # Reverse the current k-group. 
            # prev starts at group_next to seamlessly connect the tail of the
            # reversed group to the remaining part of the list.
            prev, curr = group_next, group_prev.next
            while curr != group_next:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt
            
            # The original head of the group is now the tail.
            # Connect the previous group's tail to the new head of this reversed group.
            temp = group_prev.next
            group_prev.next = kth
            
            # Move group_prev to the tail of the current reversed group
            group_prev = temp
            
        return dummy.next